### DISCUSSION TITLE
Is there any way to upload a very large file, such as 10G

### DISCUSSION THREAD

**Discussion by try1995** [pocketbase/pocketbase#6194] [Q&A] [answered]
Hey everyone,
I saw that there was a limit on the size of upload files in the code, I modifed the limit (DefaultfiledMaxsize) to 10G, but the web interface upload was still wrong. my file storage is oss.
mechine: linux, arm, 32g running memory, 256G storage space

**Comment by ganigeorgiev** [comment]
Please as mentioned in your recent opened issue, elaborate what actually is the problem or at least provide details about the returned error.

What is this "DefaultfiledMaxsize" you are referring to and how did you modify it?

**Reply by ganigeorgiev** [reply]
Do you mean that you can't set a large number in the UI file field options input?

If that's the case then it is possible to be restricted to max ~2^53 as we currently don't have special handling for BigInt values (_I'll consider it with the next release_).

**Reply by try1995** [reply]
I change the parameter value "DefaultfiledMaxsize" in the file core/field_file.go  to 10 >> 30., and "DefaultfiledMaxBodySize" to 10 >> 30 +1024 in the file apis/middlewares_body_limit.go . Then recompile and deploy. I found that 2g size can upload successfully, but no more than that.  The web interface upload will show error request. I don't konw if the interface timed out, and there were no errors in the log.
![image](https://github.com/user-attachments/assets/bdb945df-cf8d-4810-924f-fadae9498156)

**Reply by ganigeorgiev** [reply, marked as answer]
This is not what the constants are for and I don't recommend changing them to such large values. They are usually intended to be used as fallback defaults when there is no explicit value set, aka. `DefaultFileFieldMaxSize` will apply only when the related field's `MaxSize` is empty.

The problem in your case seems to be from something else. You can try to start the executable with `--dev` and check the logs in the console and/or in the _Dashboard > Logs_ section.

It could be also a JS issue, so try to check the browser's Inspect tool > Console for any errors.

While technically the max supported server-side size option value is int64, the UI "Max file size" input is a regular JS number and it is currently limited to max `Number.MAX_SAFE_INTEGER` (aka. 2^53-1) with one small caveat because of the [incorrectly used shift "trim"](https://github.com/pocketbase/pocketbase/blob/master/ui/src/components/collections/schema/SchemaFieldFile.svelte#L253) it is actually further limited to 32bit value, which I guess could explain the 2GB limit you've mentioned.

I'll consider with the next release to add proper handling for int64 values for the collection file options (or at least replace the shift with `Number` cast) but my recommendation from your earlier issue remains and if storing large files is critical for your use case you may be better using something else because PocketBase wasn't really tested or optimized for this. Uploading and serving large files over the S3 API for example could be extremely slow and resource intensive. Additionally there is also no support out of the box for chunked or partial uploads, aka. no pause/resume or progress tracking.

**Reply by try1995** [reply]
thx for answering.  I'm trying secondary developmey, if I want to change the file upload method such as support chunked or partial uploads, any suggestions

**Reply by ganigeorgiev** [reply]
Adding support for chunked or partial uploads in PocketBase would be difficult because the uploaded files are expected to be part of the record data. This allow us to handle the MIME type and size validations in a single run without having to take care for linking and cleanup in case of failed/forgotten upload attempts.

One implementation for chunked uploads that I'm aware of is https://github.com/tus/tusd but I'm not sure if this could be integrated easily with our file handling.

_To some extend you can emulate chunked uploads in PocketBase by creating a dedicated "Media/Files/Storage/etc." collection with a single `file` field + some unique identifier to associate the related chunks with one another and after the upload of the last file/chunk you can programmatically trigger "merge" for all the related files/chunks into a single file (_and eventually maybe also run the `FileField.ValidateValue(ctx, app, record)` validations for the resulting file_). But again, I'm not sure I can recommend this and you might be better using something else more suitable for your use case._

**Reply by try1995** [reply]
thx a lot, very helpful

**Reply by thangngoc89** [reply]
@try1995 I’m currently working on a similar feature but with a different approach, enabling clients to upload files directly to S3 using presigned URLs.

You can take a look at this example using Uppy (https://uppy.io/docs/aws-s3/). and the [reference server implementation in node.js](https://github.com/transloadit/uppy/blob/main/examples/aws-nodejs/index.js).

Direct uploads from the client to S3 are generally preferred for handling large files, as this reduces the memory load on the server.
